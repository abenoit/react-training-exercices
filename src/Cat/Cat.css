.Cat {
  padding: 10px 0;
  font-size: 15px;
  display: flex;
  justify-content: center;
}

.Cat button {
  margin: 0 20px;
}

.Cat .name {
  font-size: 1.5em;
  padding: 0 5px;
  color: darkcyan;
}
