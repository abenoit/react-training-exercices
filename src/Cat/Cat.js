import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './Cat.css';

const propTypes = {
  name: PropTypes.string.isRequired,
  age: PropTypes.number.isRequired,
  meowsPerHour: PropTypes.number,
  color: PropTypes.string
};

const defaultProps = {
  meowsPerHour: 10,
  color: 'deeppink'
};

class Cat extends Component {
  render() {
    return (
      <div className="Cat">
        <p>
          I am
          <span className="name" style={{ color: this.props.color }}>
            {this.props.name}
          </span>
          ! I am
          {this.props.age} years old and I do {this.props.meowsPerHour} meows
          per hour
        </p>
      </div>
    );
  }
}

Cat.propTypes = propTypes;
Cat.defaultProps = defaultProps;
export default Cat;
