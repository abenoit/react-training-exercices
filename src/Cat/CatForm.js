import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Form, ControlLabel, FormControl, Button } from 'react-bootstrap';
import './Cat.css';

const propTypes = {
  addCat: PropTypes.func.isRequired
};

const defaultCat = {
  name: '',
  age: 1,
  meowsPerHour: 10,
  color: 'green'
};

class CatForm extends Component {
  constructor() {
    super();

    this.state = {
      cat: { ...defaultCat }
    };

    this.handleChangeName = this.handleChangeName.bind(this);
    this.handleChangeAge = this.handleChangeAge.bind(this);
    this.handleChangeMeowsPerHour = this.handleChangeMeowsPerHour.bind(this);
    this.handleChangeColor = this.handleChangeColor.bind(this);
    this.addCat = this.addCat.bind(this);
  }

  handleChangeName(event) {
    this.setState({ cat: { ...this.state.cat, name: event.target.value } });
  }

  handleChangeAge(event) {
    this.setState({
      cat: { ...this.state.cat, age: event.target.valueAsNumber }
    });
  }

  handleChangeMeowsPerHour(event) {
    this.setState({
      cat: { ...this.state.cat, meowsPerHour: event.target.valueAsNumber }
    });
  }

  handleChangeColor(event) {
    this.setState({
      cat: { ...this.state.cat, color: event.target.value }
    });
  }

  isSubmitDisabled() {
    return !this.state.cat.name || !this.state.cat.age;
  }

  addCat() {
    this.props.addCat(this.state.cat);
    this.setState({
      cat: { ...defaultCat }
    });
  }

  render() {
    return (
      <div className="CatForm">
        <h2> Add a new cat </h2>
        <Form inline>
          <ControlLabel>Name * </ControlLabel>
          <FormControl
            value={this.state.cat.name}
            placeholder="Name"
            onChange={this.handleChangeName}
          />

          <ControlLabel>Age * </ControlLabel>
          <FormControl
            type="number"
            value={this.state.cat.age}
            placeholder="Age"
            onChange={this.handleChangeAge}
          />

          <ControlLabel>MeowsPerHour</ControlLabel>
          <FormControl
            type="number"
            width="200px"
            value={this.state.cat.meowsPerHour}
            placeholder="MeowsPerHour"
            onChange={this.handleChangeMeowsPerHour}
          />

          <ControlLabel>Color</ControlLabel>
          <FormControl
            width="200px"
            value={this.state.cat.color}
            placeholder="Color"
            onChange={this.handleChangeColor}
          />

          <Button
            bsStyle="primary"
            onClick={this.addCat}
            disabled={this.isSubmitDisabled()}
          >
            Add this cat
          </Button>
        </Form>
      </div>
    );
  }
}

CatForm.propTypes = propTypes;
export default CatForm;
