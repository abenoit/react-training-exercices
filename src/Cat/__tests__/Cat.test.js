import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Cat from '../Cat';

describe('Cat', () => {
  it('should display a cat', () => {
    const cat = renderer.create(<Cat name="lala" age={3} />).toJSON();
    expect(cat).toMatchSnapshot();
  });

  describe('meow', () => {
    it('should increment meows if maximum is not reached', () => {
      const spyMeow = jest.fn();
      const cat = shallow(<Cat name="lala" age={3} meow={spyMeow} />);

      cat.find('Button').simulate('click');

      expect(spyMeow).toHaveBeenCalled();
    });
  });
});
