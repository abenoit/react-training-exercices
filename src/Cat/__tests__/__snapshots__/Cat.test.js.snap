// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cat should display a cat 1`] = `
<div
  className="Cat"
>
  <div
    className="info"
  >
    I am
    <span
      className="name"
      style={
        Object {
          "color": "deeppink",
        }
      }
    >
      lala
    </span>
    ! I am
    3
     years old and I do 
    10
     meows per hour
  </div>
  <button
    className="btn btn-default"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Meow
  </button>
</div>
`;
